redbot/core/settings_caches.py

```diff
@@ -122,157 +122,145 @@ async def set_ignored_guild(self, guild: discord.Guild, set_to: bool):
 class WhitelistBlacklistManager:
     def __init__(self, config: Config):
         self._config: Config = config
-        self._cached_whitelist: Dict[Optional[int], List[int]] = {}
-        self._cached_blacklist: Dict[Optional[int], List[int]] = {}
+        self._cached_whitelist: Dict[Optional[int], Set[int]] = {}
+        self._cached_blacklist: Dict[Optional[int], Set[int]] = {}
 
-    async def get_whitelist(self, guild: Optional[discord.Guild] = None) -> List[int]:
-        ret: List[int]
+    async def get_whitelist(self, guild: Optional[discord.Guild] = None) -> Set[int]:
+        ret: Set[int]
 
         gid: Optional[int] = guild.id if guild else None
 
         if gid in self._cached_whitelist:
             ret = self._cached_whitelist[gid].copy()
         else:
             if gid is not None:
-                ret = await self._config.guild_from_id(gid).whitelist()
-                if not ret:
-                    ret = []
+                ret = set(await self._config.guild_from_id(gid).whitelist())
             else:
-                ret = await self._config.whitelist()
+                ret = set(await self._config.whitelist())
 
             self._cached_whitelist[gid] = ret.copy()
 
         return ret
 
-    async def add_to_whitelist(self, guild: Optional[discord.Guild], role_or_user: List[int]):
+    async def add_to_whitelist(self, guild: Optional[discord.Guild], role_or_user: Iterator[int]):
         gid: Optional[int] = guild.id if guild else None
         role_or_user = role_or_user or []
-        if not isinstance(role_or_user, list) or not all(
+        if not isinstance(role_or_user, (set, list)) or not all(
             isinstance(r_or_u, int) for r_or_u in role_or_user
         ):
```

Comment:
There's some conflict here - type hint says `Iterable` (well, `Iterator` but that's not the point) but you're checking if it's an instance of `list` or `set` here. I'm guessing we can just make it allow any iterable as your type hint suggests. And since we're iterating right in this `if`, we can just drop the first `isinstance` check as the `TypeError` will be raised for non-iterable anyway.
```suggestion
        if not all(isinstance(r_or_u, int) for r_or_u in role_or_user):
```
Remember to also update the message in raised `TypeError` to reflect this change.